flink-mesos/src/main/java/org/apache/flink/mesos/runtime/clusterframework/MesosApplicationMasterRunner.java

```diff
@@ -0,0 +1,618 @@
+			// and does not stay lingering around unresponsive
+
+			LOG.debug("Starting process reapers for JobManager");
+
+			actorSystem.actorOf(
+				Props.create(ProcessReaper.class, resourceMaster, LOG, ACTOR_DIED_EXIT_CODE),
+				"Mesos_Resource_Master_Process_Reaper");
+
+			actorSystem.actorOf(
+				Props.create(ProcessReaper.class, jobManager, LOG, ACTOR_DIED_EXIT_CODE),
+				"JobManager_Process_Reaper");
+		}
+		catch (Throwable t) {
+			// make sure that everything whatever ends up in the log
+			LOG.error("Mesos JobManager initialization failed", t);
+
+			if (actorSystem != null) {
+				try {
+					actorSystem.shutdown();
+				} catch (Throwable tt) {
+					LOG.error("Error shutting down actor system", tt);
+				}
+			}
+
+			if (webMonitor != null) {
+				try {
+					webMonitor.stop();
+				} catch (Throwable ignored) {
+					LOG.warn("Failed to stop the web frontend", ignored);
+				}
+			}
+
+			if(artifactServer != null) {
+				try {
+					artifactServer.stop();
+				} catch (Throwable ignored) {
+					LOG.error("Failed to stop the artifact server", ignored);
+				}
+			}
+
+			return INIT_ERROR_EXIT_CODE;
+		}
+
+		// everything started, we can wait until all is done or the process is killed
+		LOG.info("Mesos JobManager started");
+
+		// wait until everything is done
+		actorSystem.awaitTermination();
+
+		// if we get here, everything work out jolly all right, and we even exited smoothly
+		if (webMonitor != null) {
+			try {
+				webMonitor.stop();
+			} catch (Throwable t) {
+				LOG.error("Failed to stop the web frontend", t);
+			}
+		}
+
+		try {
+			artifactServer.stop();
+		} catch (Throwable t) {
+			LOG.error("Failed to stop the artifact server", t);
+		}
+
+		return 0;
+	}
+
+	// ------------------------------------------------------------------------
+	//  For testing, this allows to override the actor classes used for
+	//  JobManager and the archive of completed jobs
+	// ------------------------------------------------------------------------
+
+	protected Class<? extends MesosFlinkResourceManager> getResourceManagerClass() {
+		return MesosFlinkResourceManager.class;
+	}
+
+	protected Class<? extends JobManager> getJobManagerClass() {
+		return MesosJobManager.class;
+	}
+
+	protected Class<? extends MemoryArchivist> getArchivistClass() {
+		return MemoryArchivist.class;
+	}
+
+	protected Class<? extends TaskManager> getTaskManagerClass() {
+		return MesosTaskManager.class;
+	}
+
+	/**
+	 * Validates a condition, throwing a RuntimeException if the condition is violated.
+	 *
+	 * @param condition The condition.
+	 * @param message The message for the runtime exception, with format variables as defined by
+	 *                {@link String#format(String, Object...)}.
+	 * @param values The format arguments.
+	 */
+	private static void require(boolean condition, String message, Object... values) {
+		if (!condition) {
+			throw new RuntimeException(String.format(message, values));
+		}
+	}
```

Comment:
We could replace this via `o.a.f.util.Preconditions.checkArgument` or `checkState`.